modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RemediateProcessor.java

```diff
@@ -0,0 +1,90 @@
+ * or more contributor license agreements. Licensed under the "Elastic License
+ * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
+ * Public License v 1"; you may not use this file except in compliance with, at
+ * your election, the "Elastic License 2.0", the "GNU Affero General Public
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.ingest.common;
+
+import org.elasticsearch.cluster.metadata.ProjectId;
+import org.elasticsearch.ingest.AbstractProcessor;
+import org.elasticsearch.ingest.IngestDocument;
+import org.elasticsearch.ingest.Processor;
+
+import java.util.Map;
+
+/**
+ * Processor that remediates a failure document by:
+ * - Copying the original index (and routing if present) from ctx.document.* into the document metadata
+ * - Extracting the original source from ctx.document.source
+ * - Removing the "error" and "document" fields from the root
+ * - Restoring the original source fields at the root of the document
+ */
+public final class RemediateProcessor extends AbstractProcessor {
+
+    public static final String TYPE = "remediate";
+
+    RemediateProcessor(String tag, String description) {
+        super(tag, description);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public IngestDocument execute(IngestDocument document) throws Exception {
+        // Get the nested 'document' field, which holds the original document and metadata.
+        if (document.hasField("document") == false) {
+            throw new IllegalArgumentException("field [document] not present as part of path [document]");
+        }
+        Map<String, Object> failedDocument = (Map<String, Object>) document.getFieldValue("document", Map.class);
+
+        // Copy the original index and routing back to the document's metadata.
+        String originalIndex = (String) failedDocument.get("index");
+        if (originalIndex != null) {
+            document.setFieldValue("_index", originalIndex);
+        }
+
+        String originalRouting = (String) failedDocument.get("routing");
+        if (originalRouting != null) {
+            document.setFieldValue("_routing", originalRouting);
+        }
+
+        // Get the original document's source.
+        Map<String, Object> originalSource = (Map<String, Object>) failedDocument.get("source");
+        if (originalSource == null) {
+            throw new IllegalArgumentException("field [source] not present as part of path [document.source]");
+        }
+
+        // Remove the 'error' and 'document' fields from the top-level document.
+        document.removeField("error");
+        document.removeField("document");
+
+        // Extract all fields from the original source back to the root of the document.
+        for (Map.Entry<String, Object> entry : originalSource.entrySet()) {
+            document.setFieldValue(entry.getKey(), entry.getValue());
+        }
```

Comment:
This might work in strange ways with dotted field names. If the original source has a field named `a.b.c` then this field will create entries with `a: {b: {c: {` normalized form. This can be an issue if the original document has dotted fields that collide when normalized in this way. Since this is a implementation for ingest node provided by us, should we maybe break the glass and set the values on the document's source map directly?